src/platform/backends/shared/base_availability_zone.cpp

```diff
@@ -117,8 +117,32 @@ void BaseAvailabilityZone::set_available(const bool new_available)
     m.available = new_available;
     serialize();
 
-    for (auto& vm : m.vms)
-        vm.get().set_available(m.available);
+    try
+    {
+        for (auto& vm : m.vms)
+            vm.get().set_available(new_available);
+    }
+    catch (...)
+    {
+        // if an error occurs fallback to available.
+        m.available = true;
+        serialize();
+
+        // make sure nothing is still unavailable.
+        for (auto& vm : m.vms)
+        {
+            // setting the state here breaks encapsulation, but it's already broken.
+            std::unique_lock vm_lock{vm.get().state_mutex};
+            if (vm.get().current_state() == VirtualMachine::State::unavailable)
+            {
+                vm.get().state = VirtualMachine::State::off;
+                vm.get().update_state();
+            }
```

Comment:
Have you considered force-stop instead? That would avoid breaking encapsulation yet again and make sure there were no leftover processes.

Reply:
shutdown is ignored when an instance is unavailable. There aren't any leftover processes since an unavailable instance is always off. Using `set_available(true)` could make the VM start which is undesirable.